Added BeforeChunk and AfterChunk events

colq2 opened the pull request:
1️⃣  Why should it be added? What are the benefits of this change?

This PR adds two new events: BeforeChunk and AfterChunk.
These are triggered when exporting in chunks. With this it is possible to

- showing the progress to the user
- estimate time for an export

Related issue: #3606

2️⃣  Does it contain multiple, unrelated changes? Please separate the PRs out.

No.

3️⃣  Does it include tests, if possible?

It includes tests, but I don't know if you are quite happy with it. The test should cover, that the events are called exactly for the number of chunks. For this purpose, groups are created in the callback and counted at the end. Unfortunately I couldn't put in a callback because it can't be serialized.

If you have a better idea, let me know. I'm happy to change it.

4️⃣  Any drawbacks? Possible breaking changes?

No breaking changes.

5️⃣  Mark the following tasks as done:

- [x] Checked the codebase to ensure that your feature doesn't already exist.
- [x] Take note of the contributing guidelines.
- [x] Checked the pull requests to ensure that another person hasn't already submitted a fix.
- [x] Added tests to ensure against regression.
- [x] Updated the changelog

6️⃣  Thanks for contributing! 🙌
🙌

colq2 committed:
feat: added BeforeChunk and AfterChunk events

colq2 committed:
chore: style

colq2 committed:
chore: add changes to changelog

colq2 committed:
fix: give test a meaningful name

patrickbrouwers committed:
Merge branch '3.1' into 3.1

patrickbrouwers commented on tests/QueuedExportWithChunkEventsTest.php:
```diff
@@ -0,0 +1,45 @@
+use Maatwebsite\Excel\Tests\Data\Stubs\Database\User;
+use Maatwebsite\Excel\Tests\Data\Stubs\QueuedExportWithChunkEvents;
+
+class QueuedExportWithChunkEventsTest extends TestCase
```
Somehow this test seems to be failing in GH Actions, can you have a look?

colq2 replied:
Fixed it. Sorry, didn't saw the GH actions.

colq2 committed:
fix: wrong path for migrations in test

colq2 committed:
Merge remote-tracking branch 'origin/3.1' into 3.1

patrickbrouwers committed:
Merge branch '3.1' into 3.1

patrickbrouwers commented:
Can you already draft up a PR to the docs as well? Thanks!

colq2 committed:
feat: allow defining listener for BeforeChunk and AfterChunk with RegistersEventListeners concern

colq2 committed:
Merge remote-tracking branch 'origin/3.1' into 3.1

colq2 committed:
fix: assertDatabaseCount is undefined

colq2 commented:
Sure, added it!
While creating the PR I hove noticed, that the new events couldn't be registered with the `RegistersEventListeners` concern. So I added another test and relevant code for it.


Another point I noticed is, that all other events have some sort of context / payload e.g `Writer` or `Sheet`. It could be helpful to also add a context into BeforeChunk and AfterChunk like the `Writer` or the `AppendQueryToSheet`.
What do you think?

Djoul commented:
Hi, any reason to not merge this PR? it's a nice feature to have for all the people that want to handle the export progress. Thanks

HristoAntov commented:
+1 on merging the PR. I see no other reason for closing than inactivity.
These events are really useful for tracking current status of big exports and currenly I've found no other way to do it. Thanks

sebestenyb commented:
Can we reopen this? It is a very much missed feature.
